proposals/0010-vk-buffer-ref.md

```diff
@@ -163,6 +163,10 @@ A vk::BufferPointer can otherwise be used whereever the HLSL spec does not other
 
 Applying HLSL semantic annotations to objects of type vk::BufferPointer is disallowed.
 
+### Buffer Pointers and Type Punning Through Unions
+
+While buffer pointer types are allowed in unions, type punning with buffer pointer types is disallowed.
```

Comment:
I'm not sure this wording is sufficient.

Take this case:
```c++
Buffer<Int> I;

union Pain {
  vk::BufferPointer<SomeStruct> Ptr;
  uint64_t Int;
  double FP;
};

void Init(inout Pain P) {
  P.Int = I[0];
}

SomeStruct Load(Pain P) {
  return P.Ptr.Get();
}

void Fn() {
  Pain P;
  Init(P);
  SomeStruct S = Load(P);
}
```

The union `P` in `Fn` is initialized in `Init` to a `uint64_t`), then used as a `vk::BufferPointer`. What is the expected behavior here?

It can't be a compile error because `Init`, `Load` and `Fn` don't need to be defined in the same translation unit so we can't analyze that usage pattern. Is it undefined?

Reply:
Yes, my intention is that this would be undefined behavior.

In general, type punning across members of a union is considered undefined behavior in C++. I don't see anything in the HLSL union proposal that changes this. So it seems that this should be true no matter the types of the members, including buffer pointers.

I will augment the text of this section to be more explicit about this.